# Week 1 – Data Cleaning and Preprocessing

## Adult Income Dataset

### Objective

The objective of this project is to acquire a public dataset, investigate its data quality, identify missing values, inconsistencies, duplicate records, and unusual numerical values, and perform appropriate data cleaning and preprocessing.

The project documents the reasoning behind each preprocessing decision and evaluates the quality of the dataset before and after cleaning.

### Dataset

The dataset used in this project is the Adult dataset, which contains demographic and employment-related information and an income classification target.

### Tools Used

- Python
- Pandas
- Google Colab
- UCI Machine Learning Repository

### Main Tasks

1. Acquire the dataset.
2. Explore the dataset structure.
3. Identify missing and unknown values.
4. Investigate duplicate records.
5. Examine categorical values.
6. Investigate numerical ranges and unusual values.
7. Clean and preprocess the dataset.
8. Verify the final data quality.
9. Save the cleaned dataset.

In [ ]:
print("Week 1 Data Science Internship Project")
print("Data Acquisition, Cleaning and Preprocessing")

Week 1 Data Science Internship Project
Data Acquisition, Cleaning and Preprocessing


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

print("NumPy version:", np.__version__)
print("Pandas version:", pd.__version__)
print("Matplotlib version:", plt.matplotlib.__version__)
print("Seaborn version:", sns.__version__)

NumPy version: 2.1.3
Pandas version: 2.2.3
Matplotlib version: 3.10.0
Seaborn version: 0.13.2


In [ ]:
!pip install ucimlrepo

## 1. Dataset Acquisition

The Adult dataset was obtained from the UCI Machine Learning Repository using the `ucimlrepo` Python package.

The dataset was loaded into Pandas DataFrames containing the feature variables and target variable. The features and target were then combined into a single DataFrame for analysis.

In [ ]:
from ucimlrepo import fetch_ucirepo

adult = fetch_ucirepo(id=2)

X = adult.data.features
y = adult.data.targets

print("Dataset downloaded successfully!")
print("Features shape:", X.shape)
print("Target shape:", y.shape)

Dataset downloaded successfully!
Features shape: (48842, 14)
Target shape: (48842, 1)


## 2. Initial Dataset Exploration

The initial exploration was performed to understand the size, structure, columns, data types, and basic statistical characteristics of the dataset before applying any preprocessing.

In [ ]:
import pandas as pd

df = pd.concat([X, y], axis=1)

print("Dataset shape:", df.shape)

Dataset shape: (48842, 15)


In [ ]:
df.head()

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


In [ ]:
df.to_csv("adult_raw.csv", index=False)

print("Raw dataset saved successfully.")

Raw dataset saved successfully.


In [ ]:
import os

print(os.getcwd())

/content


In [ ]:
print(os.listdir())

['.config', 'adult_raw.csv', 'sample_data']


## 2. Initial Dataset Exploration

The initial exploration was performed to understand the size, structure, column names, data types, and basic statistical characteristics of the dataset before applying preprocessing.

In [ ]:
print("Rows and columns:")
print(df.shape)

print("\nColumn names:")
print(df.columns.tolist())

Rows and columns:
(48842, 15)

Column names:
['age', 'workclass', 'fnlwgt', 'education', 'education-num', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week', 'native-country', 'income']


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             48842 non-null  int64 
 1   workclass       47879 non-null  object
 2   fnlwgt          48842 non-null  int64 
 3   education       48842 non-null  object
 4   education-num   48842 non-null  int64 
 5   marital-status  48842 non-null  object
 6   occupation      47876 non-null  object
 7   relationship    48842 non-null  object
 8   race            48842 non-null  object
 9   sex             48842 non-null  object
 10  capital-gain    48842 non-null  int64 
 11  capital-loss    48842 non-null  int64 
 12  hours-per-week  48842 non-null  int64 
 13  native-country  48568 non-null  object
 14  income          48842 non-null  object
dtypes: int64(6), object(9)
memory usage: 5.6+ MB


In [ ]:
df.describe(include="all")

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
count,48842.000000,47879,4.884200e+04,48842,48842.000000,48842,47876,48842,48842,48842,48842.000000,48842.000000,48842.000000,48568,48842
unique,NaN,9,NaN,16,NaN,7,15,6,5,2,NaN,NaN,NaN,42,4
top,NaN,Private,NaN,HS-grad,NaN,Married-civ-spouse,Prof-specialty,Husband,White,Male,NaN,NaN,NaN,United-States,<=50K
freq,NaN,33906,NaN,15784,NaN,22379,6172,19716,41762,32650,NaN,NaN,NaN,43832,24720
mean,38.643585,NaN,1.896641e+05,NaN,10.078089,NaN,NaN,NaN,NaN,NaN,1079.067626,87.502314,40.422382,NaN,NaN
std,13.710510,NaN,1.056040e+05,NaN,2.570973,NaN,NaN,NaN,NaN,NaN,7452.019058,403.004552,12.391444,NaN,NaN
min,17.000000,NaN,1.228500e+04,NaN,1.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,1.000000,NaN,NaN
25%,28.000000,NaN,1.175505e+05,NaN,9.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,40.000000,NaN,NaN
50%,37.000000,NaN,1.781445e+05,NaN,10.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,40.000000,NaN,NaN
75%,48.000000,NaN,2.376420e+05,NaN,12.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,45.000000,NaN,NaN


### 3.2 Missing and Unknown Values

Missing values were investigated using Pandas. The dataset contained both actual missing values (`NaN`) and unknown-value placeholders represented by `?`.

The `?` placeholders were examined separately because they represent unknown information rather than valid categorical values.

In [ ]:
missing_values = df.isnull().sum()

print(missing_values)

age                 0
workclass         963
fnlwgt              0
education           0
education-num       0
marital-status      0
occupation        966
relationship        0
race                0
sex                 0
capital-gain        0
capital-loss        0
hours-per-week      0
native-country    274
income              0
dtype: int64


In [ ]:
(df == "?").sum()

,0
age,0
workclass,1836
fnlwgt,0
education,0
education-num,0
marital-status,0
occupation,1843
relationship,0
race,0
sex,0


## 3. Data Quality Assessment

The dataset was examined for common data-quality issues including missing values, placeholder values, duplicate records, inconsistent categorical labels, and unusual numerical values.

### 3.1 Data Types

The data types of all columns were inspected to distinguish numerical variables from categorical variables and to identify any potential type-related issues.

In [ ]:
print("===== DATA TYPES =====")
print(df.dtypes)

print("\n===== DUPLICATE ROWS =====")
print("Number of duplicate rows:", df.duplicated().sum())

print("\n===== UNIQUE VALUES =====")
for column in df.columns:
    print(f"{column}: {df[column].nunique()} unique values")

===== DATA TYPES =====
age                int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
income            object
dtype: object

===== DUPLICATE ROWS =====
Number of duplicate rows: 29

===== UNIQUE VALUES =====
age: 74 unique values
workclass: 9 unique values
fnlwgt: 28523 unique values
education: 16 unique values
education-num: 16 unique values
marital-status: 7 unique values
occupation: 15 unique values
relationship: 6 unique values
race: 5 unique values
sex: 2 unique values
capital-gain: 123 unique values
capital-loss: 99 unique values
hours-per-week: 96 unique values
native-country: 42 unique values
income: 4 unique values


### 3.3 Categorical Value Inspection

The unique values and frequency distributions of categorical columns were inspected to identify unexpected categories, placeholder values, and inconsistencies.

Particular attention was given to `workclass`, `occupation`, `native-country`, and the `income` target.

In [ ]:
print("===== CATEGORICAL VALUE CHECK =====")

categorical_columns = df.select_dtypes(include="object").columns

for column in categorical_columns:
    print(f"\n--- {column} ---")
    print(df[column].value_counts(dropna=False).head(20))

===== CATEGORICAL VALUE CHECK =====

--- workclass ---
workclass
Private             33906
Self-emp-not-inc     3862
Local-gov            3136
State-gov            1981
?                    1836
Self-emp-inc         1695
Federal-gov          1432
NaN                   963
Without-pay            21
Never-worked           10
Name: count, dtype: int64

--- education ---
education
HS-grad         15784
Some-college    10878
Bachelors        8025
Masters          2657
Assoc-voc        2061
11th             1812
Assoc-acdm       1601
10th             1389
7th-8th           955
Prof-school       834
9th               756
12th              657
Doctorate         594
5th-6th           509
1st-4th           247
Preschool          83
Name: count, dtype: int64

--- marital-status ---
marital-status
Married-civ-spouse       22379
Never-married            16117
Divorced                  6633
Separated                 1530
Widowed                   1518
Married-spouse-absent      628
Married-AF-spouse

### 3.4 Numerical Data Investigation

Numerical columns were examined using descriptive statistics, including minimum and maximum values, averages, and quartiles.

Potentially unusual values were investigated before deciding whether they represented actual errors. Extreme values were not automatically removed simply because they were statistically unusual.

In [ ]:
print("===== NUMERICAL SUMMARY =====")

numerical_columns = df.select_dtypes(include="number").columns

print(df[numerical_columns].describe().T)

===== NUMERICAL SUMMARY =====
                  count           mean            std      min       25%  \
age             48842.0      38.643585      13.710510     17.0      28.0   
fnlwgt          48842.0  189664.134597  105604.025423  12285.0  117550.5   
education-num   48842.0      10.078089       2.570973      1.0       9.0   
capital-gain    48842.0    1079.067626    7452.019058      0.0       0.0   
capital-loss    48842.0      87.502314     403.004552      0.0       0.0   
hours-per-week  48842.0      40.422382      12.391444      1.0      40.0   

                     50%       75%        max  
age                 37.0      48.0       90.0  
fnlwgt          178144.5  237642.0  1490400.0  
education-num       10.0      12.0       16.0  
capital-gain         0.0       0.0    99999.0  
capital-loss         0.0       0.0     4356.0  
hours-per-week      40.0      45.0       99.0  


### 3.5 Investigation of Unusual Numerical Values

Specific numerical ranges were investigated to determine whether extreme observations appeared to be invalid.

Examples included age, hours worked per week, and capital gains. Since unusual values were not automatically confirmed as errors, they were retained unless there was sufficient evidence that they were invalid.

In [ ]:
print("===== EXTREME / UNUSUAL NUMERICAL VALUES =====")

print("\nCapital-gain = 99999:")
print((df["capital-gain"] == 99999).sum())

print("\nCapital-loss > 0:")
print((df["capital-loss"] > 0).sum())

print("\nHours-per-week below 10:")
print((df["hours-per-week"] < 10).sum())

print("\nHours-per-week above 80:")
print((df["hours-per-week"] > 80).sum())

print("\nAge below 18:")
print((df["age"] < 18).sum())

print("\nAge above 80:")
print((df["age"] > 80).sum())

===== EXTREME / UNUSUAL NUMERICAL VALUES =====

Capital-gain = 99999:
244

Capital-loss > 0:
2282

Hours-per-week below 10:
700

Hours-per-week above 80:
318

Age below 18:
595

Age above 80:
148


### 3.6 Missing-Value Pattern Analysis

The number of rows containing at least one missing or unknown value was calculated to understand the overall extent of incomplete records in the dataset.

In [ ]:
print("===== MISSING VALUE PATTERNS =====")

missing_mask = (
    df["workclass"].isna() |
    df["occupation"].isna() |
    df["native-country"].isna() |
    (df["workclass"] == "?") |
    (df["occupation"] == "?") |
    (df["native-country"] == "?")
)

print("Rows containing at least one missing/unknown value:", missing_mask.sum())

print("\nRows with no missing/unknown values:", (~missing_mask).sum())

===== MISSING VALUE PATTERNS =====
Rows containing at least one missing/unknown value: 3620

Rows with no missing/unknown values: 45222


### 3.7 Detailed Duplicate Inspection

The duplicate records were inspected before removal to confirm that the identified records represented exact duplicate observations.

In [ ]:
duplicates = df[df.duplicated(keep=False)]

print("Total duplicate rows:", len(duplicates))

duplicates.head(10)

Total duplicate rows: 57


,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
2303,90,Private,52386,Some-college,10,Never-married,Other-service,Not-in-family,Asian-Pac-Islander,Male,0,0,35,United-States,<=50K
3917,19,Private,251579,Some-college,10,Never-married,Other-service,Own-child,White,Male,0,0,14,United-States,<=50K
4325,25,Private,308144,Bachelors,13,Never-married,Craft-repair,Not-in-family,White,Male,0,0,40,Mexico,<=50K
4767,21,Private,250051,Some-college,10,Never-married,Prof-specialty,Own-child,White,Female,0,0,10,United-States,<=50K
4881,25,Private,308144,Bachelors,13,Never-married,Craft-repair,Not-in-family,White,Male,0,0,40,Mexico,<=50K
4940,38,Private,207202,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,48,United-States,>50K
5104,90,Private,52386,Some-college,10,Never-married,Other-service,Not-in-family,Asian-Pac-Islander,Male,0,0,35,United-States,<=50K
5579,27,Private,255582,HS-grad,9,Never-married,Machine-op-inspct,Not-in-family,White,Female,0,0,40,United-States,<=50K
5805,20,Private,107658,Some-college,10,Never-married,Tech-support,Not-in-family,White,Female,0,0,10,United-States,<=50K
5842,25,Private,195994,1st-4th,2,Never-married,Priv-house-serv,Not-in-family,White,Female,0,0,40,Guatemala,<=50K


## 4. Data Cleaning and Preprocessing

Based on the data-quality assessment, several preprocessing steps were performed. The cleaning process addressed confirmed data-quality issues while avoiding unnecessary removal of potentially valid observations.

The main preprocessing steps were:

1. Convert `?` placeholders to missing values.
2. Remove duplicate records.
3. Standardize the income labels.
4. Remove rows containing missing values.
5. Reset the dataset index.
6. Perform final data-quality verification.

In [ ]:
# ==========================================
# STEP 16 - DATA CLEANING
# ==========================================

# Create a copy so that the original raw dataframe remains unchanged
df_clean = df.copy()

print("Original dataset shape:", df_clean.shape)

# ------------------------------------------
# 1. Convert '?' placeholders to NaN
# ------------------------------------------

df_clean = df_clean.replace("?", pd.NA)

print("\nMissing values after converting '?' to NaN:")
print(df_clean.isna().sum())

# ------------------------------------------
# 2. Remove duplicate rows
# ------------------------------------------

duplicates_before = df_clean.duplicated().sum()

df_clean = df_clean.drop_duplicates()

print("\nDuplicate rows removed:", duplicates_before)
print("Shape after removing duplicates:", df_clean.shape)

# ------------------------------------------
# 3. Standardize income labels
# ------------------------------------------

df_clean["income"] = df_clean["income"].str.replace(".", "", regex=False)

print("\nIncome categories after standardization:")
print(df_clean["income"].value_counts())

# ------------------------------------------
# 4. Remove rows containing missing values
# ------------------------------------------

missing_rows_before = df_clean.isna().any(axis=1).sum()

df_clean = df_clean.dropna()

print("\nRows removed due to missing values:", missing_rows_before)
print("Final cleaned dataset shape:", df_clean.shape)

# ------------------------------------------
# 5. Reset index
# ------------------------------------------

df_clean = df_clean.reset_index(drop=True)

print("\nFinal index reset successfully.")

print("\nFinal cleaned dataset preview:")
display(df_clean.head())

Original dataset shape: (48842, 15)

Missing values after converting '?' to NaN:
age                  0
workclass         2799
fnlwgt               0
education            0
education-num        0
marital-status       0
occupation        2809
relationship         0
race                 0
sex                  0
capital-gain         0
capital-loss         0
hours-per-week       0
native-country     857
income               0
dtype: int64

Duplicate rows removed: 29
Shape after removing duplicates: (48813, 15)

Income categories after standardization:
income
<=50K    37128
>50K     11685
Name: count, dtype: int64

Rows removed due to missing values: 3619
Final cleaned dataset shape: (45194, 15)

Final index reset successfully.

Final cleaned dataset preview:


,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


In [ ]:
# Save the cleaned dataset

df_clean.to_csv("adult_cleaned.csv", index=False)

print("Cleaned dataset saved successfully.")
print("File name: adult_cleaned.csv")
print("Final shape:", df_clean.shape)

Cleaned dataset saved successfully.
File name: adult_cleaned.csv
Final shape: (45194, 15)


In [ ]:
print("===== CLEANING VERIFICATION =====")

print("\nFinal shape:")
print(df_clean.shape)

print("\nMissing values:")
print(df_clean.isna().sum())

print("\nDuplicate rows:")
print(df_clean.duplicated().sum())

print("\nIncome categories:")
print(df_clean["income"].value_counts())

print("\nUnique income labels:")
print(df_clean["income"].unique())

===== CLEANING VERIFICATION =====

Final shape:
(45194, 15)

Missing values:
age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64

Duplicate rows:
19

Income categories:
income
<=50K    33988
>50K     11206
Name: count, dtype: int64

Unique income labels:
['<=50K' '>50K']


### 4.1 Final Duplicate Check

After standardizing the income labels, a second duplicate check was performed.

This was necessary because two records that previously had different textual income labels could become identical after removing the trailing period.

The remaining duplicate records were therefore removed before the final dataset was saved.

In [ ]:
# Remove duplicates created after income standardization

duplicates_remaining = df_clean.duplicated().sum()

print("Duplicates before final duplicate removal:", duplicates_remaining)

df_clean = df_clean.drop_duplicates().reset_index(drop=True)

print("Duplicates removed:", duplicates_remaining)
print("Final cleaned dataset shape:", df_clean.shape)

Duplicates before final duplicate removal: 19
Duplicates removed: 19
Final cleaned dataset shape: (45175, 15)


## 5. Final Data Quality Verification

The cleaned dataset was checked again after preprocessing.

The final verification examined:

- Dataset dimensions
- Missing values
- Duplicate records
- Income categories
- Unique income labels

The purpose of this step was to confirm that the cleaning operations produced the expected final dataset.

In [ ]:
print("===== FINAL DATA QUALITY CHECK =====")

print("\nFinal shape:")
print(df_clean.shape)

print("\nMissing values:")
print(df_clean.isna().sum().sum())

print("\nDuplicate rows:")
print(df_clean.duplicated().sum())

print("\nIncome categories:")
print(df_clean["income"].value_counts())

print("\nUnique income labels:")
print(df_clean["income"].unique())

===== FINAL DATA QUALITY CHECK =====

Final shape:
(45175, 15)

Missing values:
0

Duplicate rows:
0

Income categories:
income
<=50K    33973
>50K     11202
Name: count, dtype: int64

Unique income labels:
['<=50K' '>50K']


## 6. Save the Cleaned Dataset

After the final quality checks confirmed that the dataset contained no missing values or duplicate records, the cleaned dataset was saved as `adult_cleaned.csv`.

This file will be used as the cleaned dataset for subsequent analysis and modeling tasks.

In [ ]:
df_clean.to_csv("adult_cleaned.csv", index=False)

print("Final cleaned dataset saved successfully.")
print("Shape:", df_clean.shape)

Final cleaned dataset saved successfully.
Shape: (45175, 15)


In [ ]:
from google.colab import files

files.download("adult_raw.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
files.download("adult_cleaned.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## 7. Conclusion

The Adult dataset was successfully investigated and cleaned using Python and Pandas.

The preprocessing process addressed missing and unknown values, duplicate records, and inconsistent income labels. Numerical values were investigated for unusual ranges, but potentially valid extreme observations were not removed without sufficient evidence that they represented errors.

The final dataset contains 45,175 rows and 15 columns. The final verification confirmed that the cleaned dataset contains no missing values or duplicate rows, and the income target contains two standardized categories: `<=50K` and `>50K`.

The cleaned dataset was saved as `adult_cleaned.csv` for use in subsequent analysis and modeling.